app_dart/lib/src/request_handlers/push_gold_status_to_github.dart

```diff
@@ -0,0 +1,274 @@
+    AuthenticationProvider authenticationProvider, {
+    @visibleForTesting DatastoreServiceProvider datastoreProvider,
+    @visibleForTesting LoggingProvider loggingProvider,
+    HttpClient goldClient,
+  })  : datastoreProvider =
+            datastoreProvider ?? DatastoreService.defaultProvider,
+        loggingProvider = loggingProvider ?? Providers.serviceScopeLogger,
+        goldClient = goldClient ?? HttpClient(),
+        super(config: config, authenticationProvider: authenticationProvider);
+
+  final DatastoreServiceProvider datastoreProvider;
+  final LoggingProvider loggingProvider;
+  final HttpClient goldClient;
+
+  static const String kTargetUrl = 'https://flutter-gold.skia.org/changelists';
+  static const String kContext = 'flutter-gold';
+
+  @override
+  Future<Body> get() async {
+    final Logging log = loggingProvider();
+    final DatastoreService datastore = datastoreProvider();
+
+    if (authContext.clientContext.isDevelopmentEnvironment) {
+      // Don't push gold status from the local dev server.
+      return Body.empty;
+    }
+
+    const RepositorySlug slug = RepositorySlug('flutter', 'flutter');
+    final GitHub gitHubClient = await config.createGitHubClient();
+    final GraphQLClient cirrusClient = await config.createCirrusGraphQLClient();
+    final List<GithubGoldStatusUpdate> statusUpdates =
+        <GithubGoldStatusUpdate>[];
+    final List<String> cirrusCheckStatuses = <String>[];
+    const List<String> cirrusInProgressStates = <String>[
+      'EXECUTING',
+      'CREATED',
+      'TRIGGERED',
+      'NEEDS_APPROVAL'
+    ];
+
+    await for (PullRequest pr in gitHubClient.pullRequests.list(slug)) {
+      // Check run statuses for this pr
+      final List<dynamic> cirrusStatuses =
+          await queryCirrusGraphQL(pr.head.sha, cirrusClient, log, 'flutter');
+      for (dynamic runStatus in cirrusStatuses) {
+        final String status = runStatus['status'];
+        final String taskName = runStatus['name'];
+
+        log.debug(
+            'Found Cirrus build status for pull request ${pr.number}, commit '
+            '${pr.head.sha}: $taskName ($status)');
+
+        if (taskName.contains('framework')) {
+          // Analyzer requires curly braces here.
+          cirrusCheckStatuses.add(status);
+        }
+      }
+
+      // Is there a framework test? (Generates gold tryjobs)
+      if (cirrusCheckStatuses.isEmpty) {
+        // Analyzer requires curly braces here.
+        return Body.empty;
+      }
+
+      // This PR needs a Gold Status
+      // Get last known Gold status from datastore
+      final GithubGoldStatusUpdate lastUpdate =
+          await datastore.queryLastGoldUpdate(slug, pr);
+      CreateStatus statusRequest;
+
+      log.debug('Last known Gold status for ${pr.number} was with sha '
+          '${lastUpdate?.head ?? 'initial'}, status: ${lastUpdate?.status ?? 'initial'}');
+
+      if (lastUpdate.head == null || lastUpdate.head != pr.head.sha) {
```

Comment:
nit: This whole if/else block would read more clearly if it was broken up into a few functions.

Reply:
I have refactored, I think it is much simpler now. :)